bundles/org.openhab.binding.insteon/src/main/java/org/openhab/binding/insteon/internal/device/DeviceFeature.java

```diff
@@ -0,0 +1,428 @@
+ * 4) PollHandler: creates an Insteon message to query the DeviceFeature
+ *
+ * Lastly, DeviceFeatureListeners can register with the DeviceFeature to get notifications when
+ * the state of a feature has changed. In practice, a DeviceFeatureListener corresponds to an
+ * OpenHAB item.
+ *
+ * The character of a DeviceFeature is thus given by a set of message and command handlers.
+ * A FeatureTemplate captures exactly that: it says what set of handlers make up a DeviceFeature.
+ *
+ * DeviceFeatures are added to a new device by referencing a FeatureTemplate (defined in device_features.xml)
+ * from the Device definition file (device_types.xml).
+ *
+ * @author Daniel Pfrommer - Initial contribution
+ * @author Bernd Pfrommer - OpenHAB 1 insteonplm binding
+ * @author Rob Nielsen - Port to OpenHAB 2 insteon binding
+ */
+@NonNullByDefault
+public class DeviceFeature {
+    public static enum QueryStatus {
+        NEVER_QUERIED,
+        QUERY_PENDING,
+        QUERY_ANSWERED
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(DeviceFeature.class);
+
+    private static HashMap<String, FeatureTemplate> s_features = new HashMap<String, FeatureTemplate>();
+
+    private InsteonDevice m_device = new InsteonDevice();
+    private String m_name = "INVALID_FEATURE_NAME";
+    private boolean m_isStatus = false;
+    private int m_directAckTimeout = 6000;
+    private QueryStatus m_queryStatus = QueryStatus.NEVER_QUERIED;
+
+    private @Nullable MessageHandler m_defaultMsgHandler = new MessageHandler.DefaultMsgHandler(this);
+    private @Nullable CommandHandler m_defaultCommandHandler = new CommandHandler.WarnCommandHandler(this);
+    private @Nullable PollHandler m_pollHandler = null;
+    private @Nullable MessageDispatcher m_dispatcher = null;
+
+    private HashMap<Integer, @Nullable MessageHandler> m_msgHandlers = new HashMap<Integer, @Nullable MessageHandler>();
+    private HashMap<Class<? extends Command>, @Nullable CommandHandler> m_commandHandlers = new HashMap<Class<? extends Command>, @Nullable CommandHandler>();
+    private ArrayList<DeviceFeatureListener> m_listeners = new ArrayList<DeviceFeatureListener>();
+    private ArrayList<DeviceFeature> m_connectedFeatures = new ArrayList<DeviceFeature>();
+
+    /**
+     * Constructor
+     *
+     * @param device Insteon device to which this feature belongs
+     * @param name descriptive name for that feature
+     */
+    public DeviceFeature(InsteonDevice device, String name) {
+        m_name = name;
+        setDevice(device);
+    }
+
+    /**
+     * Constructor
+     *
+     * @param name descriptive name of the feature
+     */
+    public DeviceFeature(String name) {
+        m_name = name;
+    }
+
+    // various simple getters
+    public String getName() {
+        return m_name;
+    }
+
+    public synchronized QueryStatus getQueryStatus() {
+        return m_queryStatus;
```

Comment:
The getter is also synchronized, this is necessary.

Reply:
synchronization is only useful when attempting to executing multiple statements of code in an atomic fashion. The code above is already atomic just by virtue of it being a return statement without an any addition logic.
A proper way to handle this would be to make `m_queryStatus` `volatile` and remove `synchronized` from the getter and setter.

Reply:
I'm going to leave this as it. I has been this way forever.

Reply:
That is fine, it is harmless the way it is.